### PR TITLE
add best practice rule href-no-hash

### PR DESCRIPTION
**Pull request by badtant**
New best practice rule as discussed here: https://github.com/dequelabs/axe-core/issues/209

### CODE REVIEW

**Comment by badtant** [comment]
It says "Contributor License Agreement is not signed yet" but when I click on Details it says "You have signed the CLA for dequelabs/axe-core" ??

**Comment by dylanb** [comment]
@badtant despite what @marcysutton said in the issue referenced, I don't think this rule meets the criteria for a best practice axe-core rule.

In order for a rule to be included it must never flag things that are not problems. In this case many uses of an anchor with just a # as an href (and the accompanying JavaScript) will not be problems.

The difference between a best practice and a violation is that best practices are issues that can be found with 100% certainty but where the application would still be considered compliant with the issue (WCAG or Section 508). An example is: WCAG requires one mechanism to bypass repetitive navigation. If we don't find a mechanism at all that is a strict WCAG violation. If we find one mechanism but not all the mechanisms (e.g. a skip link, headings and landmark regions) then we flag this as a best practice.

Does that make sense?

**Comment by badtant** [comment]
Hmm, ok. Don't quite understand...
If someone uses a link with only # as href. That link wont lead anywhere... so it shouldn't be a link? Therefor best practise should be to use a button instead so that the screen reader users don't get confused by being told it's a link. In my opinion that is a problem.

Or is there some case that I'm missing?

**Comment by dylanb** [comment]
It depends on what is happening. If the anchor is acting on something e.g. deleting an item, popping up a dialog etc. then it should definitely be a button. If however it is navigating to some other portion of the app/page then it is ok for it to be a link. We cannot tell, so we there would be a high percentage of false positives

**Comment by badtant** [comment]
But in those cases, shouldn't there be a url or hash+id as href and not just a hash?

**Comment by dylanb** [comment]
If the app is written using a router, then it will have proper hashes, if not, then it depends on how its implemented. Take a look at jQuery mobile

http://demos.jquerymobile.com/1.3.0/docs/widgets/navbar/

It uses a LOT of anchors with # in its examples and anyone using that framework would get tons of false positives. This is just one example.

**Comment by badtant** [comment]
And in my opinion that is an accessibility issue. Just because a framework promotes some code it doesn't mean that it's accessible. And also, the rule in the PR is disabled by default so only advanced users that put effort to configuration is affected.

**Comment by Ringen** [comment]
I can see the problem with false positives but, as @badtant mentioned, if the rule is disabled by default and you have to manually enable it I don't think it's an issue.

**Comment by dylanb** [comment]
@Ringen yes that is true, however it does go against our manifesto

**Comment by badtant** [comment]
For example it is highlighted as a bad example here:
http://webaim.org/techniques/hypertext/

So I would argue that it isn't a false positive, just bad code :)

**Comment by dylanb** [comment]
@badtant that is a bad example because of its use of `onmouseover` not because of the use of the `#` in the `href`

**Comment by badtant** [comment]
"One of the most serious barriers is to create links that go nowhere."

"The link in this example goes nowhere. Its only purpose is to activate a JavaScript function."

I stil think it's best practice to use buttons instead. And was hopeing that Axe would promote that in order to get libraries such as jquery mobile (mentioned above) aware of accessibility issues that they are promoting. 

But I'm gonna stop argue now :)

**Comment by dylanb** [comment]
ok, I think we could create a useful best practice rule if we clarify what we are pointing out a bit more precisely.

The best practice that is 100% never a false positive is that anchor links should be fragments or well formed URIs. I think this is what you are trying to achieve - correct?

If that is the case, the messages should be clearer and we should probably be checking for well-formed URIs when the anchor is not a fragment.

I will put message suggestions as comments in the diff

File: doc/rule-descriptions.md [hunk 1]
```diff
@@ -22,6 +22,7 @@
 | empty-heading | Ensures headings have discernible text | best-practice |
 | frame-title | Ensures &lt;iframe&gt; and &lt;frame&gt; elements contain a unique and non-empty title attribute | wcag2a, wcag241, section508, section508.22.i |
 | heading-order | Ensures the order of headings is semantically correct | best-practice |
+| href-no-hash | Ensures an anchors href value is not just # | best-practice |
```
Comment (dylanb) [inline doc/rule-descriptions.md]: How about "Ensures that href values are valid link references to promote only using anchors as links"

File: lib/rules/href-no-hash.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
+  "id": "href-no-hash",
+  "selector": "a[href]",
+  "enabled": false,
+  "tags": [
+    "best-practice"
+  ],
+  "metadata": {
+    "description": "Ensures an anchors href value is not just #",
+    "help": "Anchors with href value that is just # should probably be a button"
```
Comment (dylanb) [inline lib/rules/href-no-hash.json]: How about the help reading: "Anchors must only be used as links and must therefore have an href value that is a valid reference"

**Comment by badtant** [comment]
Yes, that is what I'm trying to achieve :)
I've updated the texts.

**Comment by dylanb** [comment]
Messages are good.

What about checking for well formed URIs as well?

**Comment by badtant** [comment]
That's going to be quite tricky. Maybe should look for some third party code for that?

**Comment by dylanb** [comment]
A fully generalized implementation is quite complex but I think we might be able to do a simple implementation using regexes that would cover most of what people would want to put into an anchor (for example, we probably don't need to handle `username:password@`)

If you accept the license again, I'll merge this PR, we can open another issue to improve it to do URI checking.

**Comment by badtant** [comment]
When I click the details link I get the message "You have signed the CLA for dequelabs/axe-core" so don't know what more to sign.

**Comment by dylanb** [comment]
click the details link next to the error above. It should take you to the CLA.

**Comment by badtant** [comment]
When I click it it says "You have signed the CLA for dequelabs/axe-core"

**Comment by badtant** [comment]
Finally got rid of the message :)

**Comment by dylanb** [comment]
one more thing, could you add a test with an anchor that has no href at all?

**Comment by badtant** [comment]
I've added a test for an empty href. Since the selector only looks for anchors with a href attribute there is no need for tests with anchors without href.

I've also updated rules-description.md with a column that shows which rules are enabled/disabled by default.

**Comment by badtant** [comment]
Hmm, something happened with the Node 4 tests. Can it be run again?

**Pull request merged**